### DISCUSSION TITLE
Port NextJS

### DISCUSSION THREAD

**Discussion by zzau13** [zzau13/yarte#236] [Ideas]
This is the place to talk about this madness.

**Comment by Chaostheorie** [comment]
List of things I think are important/ required (open for discussion):

- A usable frontend library that has interop (for at least types) with backend
    - Note: This won't happen without JS as output, since it's the current way to got and WASM will not take over this role in the next *decades*as far as I can tell, with e.g., *Nim*, Reason, Dart, Elm … being way more adoptable and already ready for such a job
- Web-framework (maybe actix-web, tide …) with integration for frontend library (e.g., SSR/ Type and method interop)
- A good/ usable build system with:
    -  tools for common tasks, such as asset optimization
    - friendly approach to config
    - customization
    - Fast build times (HMR for frontend)
    - Support for existing tooling (e.g. SASS, TS …) since adoption will at best happen by migration and over time
- Support for other common features: localization, preview mode …

While researching on this topic I came across [Seed](https://seed-rs.org/) a rust framework for creating web applications that builds, as far as I could tell, around the MVC model. I've also found [yew](https://yew.rs/), in conjunction with [trunk](https://github.com/thedodd/trunk), to be very usable for small applications, though yew is not yet production ready.

I'm currently working on a POC for a yew based JAMstack with actix-web as backend. Furthermore, I will also try to produce an example for alternative backend, since actix-web may seem unintuitive if you aren't used to actors.

A quick note on my current idea for a 'pure'-rust stack:

- Build system: cargo-make + optimization tools like svgo (can be reconciled later on but no need to reinvent the wheel)
- Backend: Either Actix-web, Warp or Rocket (doesn't really matter, but those have at least a mature ecosystem)
- Frontend: Either yew, seed, or something else. 
- DB: Depends on backend. Would recommend sqlx and then either PostgreSQL or MySQL, but that more or less application specific anyway.

Sharing code can be done by having an intermediate shared lib that exposes types, structs, and logic for backend and frontend. I'm trying to make it build with docker+docker-compose to ensure portability.
